### PR TITLE
[MCI] Ingestion layer changes

### PR DESCRIPTION
**Pull request by swathinsankaran**
This PR contains the following changes:

1. Moves the new CRDs MultiClusterIngress, ClusterSet, and ServiceImport to AKO's CRD folder.
2. Flag to enable/disable MultiClusterIngress feature in values.yaml
3. Defines the cluster roles for the new APIs.
4. Ingestion layer changes to process the new CRDs.
5. Status layer changes to update the validation errors in the status.

TODO: validation code and unit tests

### CODE REVIEW

File: internal/k8s/crdcontroller.go [hunk 1]
```diff
@@ -499,7 +504,136 @@ func (c *AviController) SetupIstioCRDEventHandlers(numWorkers uint32) {
...
+	serviceImportEventHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			if c.DisableSync {
+				return
+			}
+			si := obj.(*akov1alpha1.ServiceImport)
+			namespace, _, _ := cache.SplitMetaNamespaceKey(utils.ObjKey(si))
+			key := lib.ServiceImport + "/" + utils.ObjKey(si)
+			if err := validateServiceImportObj(key, si); err != nil {
+				utils.AviLog.Warnf("key: %s, msg: Validation of ServiceImport failed: %v", key, err)
```
Comment (monotosh-avi) [inline internal/k8s/crdcontroller.go:587]: Should we add the key if validation fails ?

**Review by monotosh-avi — COMMENTED** [review]
Looks good to me Swathin, one comment.

File: internal/k8s/crdcontroller.go (hunk 1 above)
Comment (swathinsankaran) [reply internal/k8s/crdcontroller.go:583]: I have added the code to stop execution when the validation fails.

**Review by akshayhavile — COMMENTED** [review]
@swathinsankaran @monotosh-avi : I also would like to understand how to CRDs are mapped each other to create single object. I couldn't figure out single common property between those.

File: cmd/ako-main/main.go [hunk 2]
```diff
@@ -27,6 +27,7 @@ import (
 	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/internal/lib"
 	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pkg/api"
 	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pkg/api/models"
+	akocrd "github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pkg/client/v1alpha1/clientset/versioned"
```
[the remark this answers is not in the document]
Comment (swathinsankaran) [reply cmd/ako-main/main.go]: I have removed the duplicate import and used `crd` as suggested in the comment.

File: cmd/ako-main/main.go [hunk 3]
```diff
@@ -117,6 +119,13 @@ func InitializeAKC() {
 			}
 			akoControlConfig.SetIstioClientset(istioClient)
 		}
+
+		if utils.IsMultiClusterIngressEnabled() {
```
[the remark this answers is not in the document]
Comment (swathinsankaran) [reply cmd/ako-main/main.go]: I have removed the code and reused the `clientSet` created before.

File: helm/ako/crds/ako.vmware.com_multiclusteringresses.yaml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                    weight:
+                      description: "weight of this member in the resultant virtual service"
+                      type: "integer"
+                      minimum: 1
+                      maximum: 100
+          status:
+            type: object
+            properties:
+              status:
+                description: "represents whether the MCI object was accepted/rejected and the reason for rejection"
```
[the remark this answers is not in the document]
Comment (swathinsankaran) [reply helm/ako/crds/ako.vmware.com_multiclusteringresses.yaml]: I have modified the description.

File: internal/k8s/crdcontroller.go [hunk 5]
```diff
@@ -499,7 +499,138 @@ func (c *AviController) SetupIstioCRDEventHandlers(numWorkers uint32) {
...
+func (c *AviController) SetupServiceImportEventHandlers(numWorkers uint32) {
+	utils.AviLog.Infof("Setting up ServiceImport CRD Event handlers")
+
+	serviceImportEventHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			if c.DisableSync {
+				return
+			}
+			si := obj.(*akov1alpha1.ServiceImport)
+			namespace, _, _ := cache.SplitMetaNamespaceKey(utils.ObjKey(si))
```
[the remark this answers is not in the document]
Comment (swathinsankaran) [reply internal/k8s/crdcontroller.go:592]: I have added the code for namespace sync.

**Review by chauhanshubham — APPROVED** [review]

File: internal/k8s/crdcontroller.go [hunk 6]
```diff
@@ -499,7 +499,138 @@ func (c *AviController) SetupIstioCRDEventHandlers(numWorkers uint32) {
...
+func (c *AviController) SetupMultiClusterIngressEventHandlers(numWorkers uint32) {
+	utils.AviLog.Infof("Setting up MultiClusterIngress CRD Event handlers")
+
+	multiClusterIngressEventHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			if c.DisableSync {
+				return
+			}
+			mci := obj.(*akov1alpha1.MultiClusterIngress)
+			namespace, _, _ := cache.SplitMetaNamespaceKey(utils.ObjKey(mci))
```
[the remark this answers is not in the document]
Comment (swathinsankaran) [reply internal/k8s/crdcontroller.go:514]: I have added the code for namespace sync as discussed.

**Review by akshayhavile — APPROVED** [review]

**Pull request merged**